### PR TITLE
Add feature flag for village type ahead search

### PR DESCRIPTION
**Pull request by msasikanth**

### CODE REVIEW

File: app/src/main/java/org/simple/clinic/feature/Feature.kt [hunk 1]
```diff
@@ -23,5 +23,6 @@ enum class Feature(
   ChangeLanguage(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M),
   MLKitQrCodeScanner(true, "ml_kit_qr_code_scanner"),
   InstantSearch(true, "instant_search"),
-  OverdueCount(true, "overdue_count")
+  OverdueCount(true, "overdue_count"),
+  VillageTypeAhead(true, "village_type_ahead")
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/feature/Feature.kt:27]: Should the default value for this be set to `false` instead, given that we'll need to wait for the remote config sync to complete.

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt [hunk 2]
```diff
@@ -21,7 +24,7 @@ class EditPatientInit(
         LoadInputFields,
     )
 
-    if (!model.hasColonyOrVillagesList) {
+    if (!model.hasColonyOrVillagesList && features.isEnabled(Feature.VillageTypeAhead)) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt:27]: We are introducing an impurity here. Can we instead pass the flag in as a `Boolean` constructor property instead?

File: app/src/test/java/org/simple/clinic/editpatient/EditPatientInitTest.kt [hunk 3]
```diff
@@ -17,7 +21,16 @@ class EditPatientInitTest {
   private val patientPhoneNumber = TestData.patientPhoneNumber(UUID.fromString("f204b770-83c5-4145-9ca7-c2273be2bbdc"), number = "9999999999")
   private val bangladeshNationalId = TestData.businessId(UUID.fromString("c9be5e5d-770c-4f37-a9de-2c304dfebfcd"))
   private val dateOfBirthFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy", Locale.ENGLISH)
-  private val initSpec = InitSpec(EditPatientInit(patient, patientAddress, patientPhoneNumber, bangladeshNationalId))
+  private val features = Features(
+      remoteConfigService = NoOpRemoteConfigService(DefaultValueConfigReader()),
+      overrides = mapOf(
+          Feature.VillageTypeAhead to true
+      ))
```
Comment (vinaysshenoy) [inline app/src/test/java/org/simple/clinic/editpatient/EditPatientInitTest.kt:28]: Yeah, as mentioned earlier, this information is unnecessary for testing business logic. Passing in the value as a `Boolean` property to `EditPatientInitTest` should simplify this a lot.

File: app/src/main/java/org/simple/clinic/newentry/PatientEntryInit.kt [hunk 4]
```diff
@@ -3,12 +3,14 @@ package org.simple.clinic.newentry
 import com.spotify.mobius.First
 import com.spotify.mobius.First.first
 import com.spotify.mobius.Init
+import org.simple.clinic.feature.Feature
+import org.simple.clinic.feature.Features
 
-class PatientEntryInit : Init<PatientEntryModel, PatientEntryEffect> {
+class PatientEntryInit(private val features: Features) : Init<PatientEntryModel, PatientEntryEffect> {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/newentry/PatientEntryInit.kt:9]: Same comments as earlier about the `EditPatientScreen`.

File: app/src/main/java/org/simple/clinic/newentry/PatientEntryInit.kt [hunk 5]
```diff
@@ -3,12 +3,14 @@ package org.simple.clinic.newentry
...
   override fun init(model: PatientEntryModel): First<PatientEntryModel, PatientEntryEffect> {
     val effects = mutableSetOf(FetchPatientEntry, LoadInputFields)
 
-    if (!model.hasColonyOrVillagesList) {
+    if (!model.hasColonyOrVillagesList && features.isEnabled(Feature.VillageTypeAhead)) {
```
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/newentry/PatientEntryInit.kt:13]: Can we also verify this behaviour in tests?

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt (hunk 2 above)
Comment (vinaysshenoy) [inline app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt:27]: Can we verify this behaviour in tests?

**Review by vinaysshenoy — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/simple/clinic/feature/Feature.kt (hunk 1 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/feature/Feature.kt:27]: Ah, good point. Will make the change.

File: app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt (hunk 2 above)
Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt:27]: We test feature flags in our tests as well? 🤔

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt]: Depends on the feature flag itself, I guess.

Comment (msasikanth) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt]: I pushed other changes, but if you can confirm if this is something we need for this feature flag I am gonna add this case and push again or else the rest of the changes are done.

Comment (vinaysshenoy) [reply app/src/main/java/org/simple/clinic/editpatient/EditPatientInit.kt]: Let's skip this for now and add the test in another PR.

**Review by vinaysshenoy — APPROVED** [review]

**Pull request merged**